flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/sessionjob/SessionJobObserver.java

```diff
@@ -17,13 +17,175 @@
+        this.defaultConfig = defaultConfig;
+    }
+
     @Override
-    public void observe(FlinkSessionJob flinkSessionJob, Context context) {}
+    public void observe(FlinkSessionJob flinkSessionJob, Context context) {
+        var lastReconciledSpec =
+                flinkSessionJob.getStatus().getReconciliationStatus().getLastReconciledSpec();
+
+        if (lastReconciledSpec == null) {
+            return;
+        }
+
+        var flinkDepOpt =
+                OperatorUtils.getSecondaryResource(flinkSessionJob, context, operatorConfiguration);
+
+        var helper = new SessionJobHelper(flinkSessionJob, LOG);
+
+        if (!helper.sessionClusterReady(flinkDepOpt)) {
+            return;
+        }
+
+        Configuration lastValidatedConfig =
+                FlinkUtils.getEffectiveConfig(flinkDepOpt.get(), this.defaultConfig);
+        var jobFound = observeFlinkJobStatus(flinkSessionJob, lastValidatedConfig);
+        if (jobFound) {
+            observeSavepointStatus(flinkSessionJob, lastValidatedConfig);
+        }
+    }
+
+    /**
+     * Observe the status of the flink job.
+     *
+     * @param sessionJob the target session job.
+     * @param lastValidatedConfig The last validated config.
+     * @return If job found return true, otherwise return false.
+     */
+    private boolean observeFlinkJobStatus(
+            FlinkSessionJob sessionJob, Configuration lastValidatedConfig) {
+        var jobId = sessionJob.getStatus().getJobStatus().getJobId();
+        var previousJobStatus = sessionJob.getStatus().getJobStatus().getState();
+        try {
+            var jobs = flinkService.listJobs(lastValidatedConfig);
+            var clusterJobStatuses =
+                    jobs.stream()
+                            .filter(job -> job.getJobId().toHexString().equals(jobId))
+                            .collect(Collectors.toList());
+            if (!clusterJobStatuses.isEmpty()) {
+                String targetJobStatus =
+                        updateJobStatus(sessionJob.getStatus().getJobStatus(), clusterJobStatuses);
+                if (targetJobStatus.equals(previousJobStatus)) {
+                    LOG.info("Job status ({}) unchanged", previousJobStatus);
+                } else {
+                    LOG.info(
+                            "Job status successfully updated from {} to {}",
+                            previousJobStatus,
+                            targetJobStatus);
+                }
+                return true;
+            } else {
+                LOG.info("No job found for JobID: {}", jobId);
+                sessionJob.getStatus().getJobStatus().setState(JOB_STATE_UNKNOWN);
+            }
+
+        } catch (Exception e) {
+            LOG.error("Exception while listing jobs", e);
+            sessionJob.getStatus().getJobStatus().setState(JOB_STATE_UNKNOWN);
+            return false;
+        }
+        return true;
+    }
+
+    private void observeSavepointStatus(
```

Comment:
Seems like there is a lot of code duplication here with the ApplicationObserver, I wonder if we can factor this out somehow

Reply:
Yes, I have the same feeling, I will try to improve this